api/v1alpha2/resource_error.go

```diff
@@ -19,70 +19,150 @@
 
 package v1alpha2
 
+import (
+	"fmt"
+	"strings"
+
+	"github.com/go-logr/logr"
+)
+
+type ResourceErrorSeverity string
+type ResourceErrorType string
+
+const (
+	SeverityMinor ResourceErrorSeverity = "Minor"
+	SeverityMajor ResourceErrorSeverity = "Major"
+	SeverityFatal ResourceErrorSeverity = "Fatal"
+)
+
+const (
+	TypeInternal ResourceErrorType = "Internal"
+	TypeUser     ResourceErrorType = "User"
+)
+
 type ResourceErrorInfo struct {
 	// Optional user facing message if the error is relevant to an end user
 	UserMessage string `json:"userMessage,omitempty"`
 
 	// Internal debug message for the error
 	DebugMessage string `json:"debugMessage"`
 
-	// Indication if the error is likely recoverable or not
-	Recoverable bool `json:"recoverable"`
+	// Internal or user error
+	// +kubebuilder:validation:Enum=Internal;User
+	Type ResourceErrorType `json:"type"`
+
+	// Indication of how sever the error is. Minor will likely succeed, Major may
+	// succeed, and Fatal will never succeed.
+	// +kubebuilder:validation:Enum=Minor;Major;Fatal
+	Severity ResourceErrorSeverity `json:"severity"`
 }
 
 type ResourceError struct {
 	// Error information
 	Error *ResourceErrorInfo `json:"error,omitempty"`
 }
 
-func NewResourceError(message string, err error) *ResourceErrorInfo {
-	resourceError := &ResourceErrorInfo{
-		Recoverable: true,
+func NewResourceError(format string, a ...any) *ResourceErrorInfo {
+	return &ResourceErrorInfo{
+		Type:         TypeInternal,
+		Severity:     SeverityMinor,
+		DebugMessage: fmt.Sprintf(format, a...),
 	}
+}
 
-	if err != nil {
-		// If the error provided is already a ResourceError, use it and concatenate
-		// the debug messages
-		_, ok := err.(*ResourceErrorInfo)
-		if ok {
-			resourceError = err.(*ResourceErrorInfo)
-		}
+func (e *ResourceErrorInfo) WithUserMessage(format string, a ...any) *ResourceErrorInfo {
+	// Only set the user message if it's empty. This prevents upper layers
+	// from overriding a user message set by a lower layer
+	if e.UserMessage == "" {
+		e.UserMessage = fmt.Sprintf(format, a...)
+	}
+
+	return e
+}
+
+func (e *ResourceErrorInfo) WithError(err error) *ResourceErrorInfo {
+	debugMessageList := []string{}
 
-		if message == "" {
-			message = err.Error()
+	childError, ok := err.(*ResourceErrorInfo)
+	if ok {
+		// Inherit the severity and the user message if the child error is a ResourceError
+		e.Severity = childError.Severity
+		e.UserMessage = childError.UserMessage
+		e.Type = childError.Type
+
+		if childError.DebugMessage == "" {
+			debugMessageList = append(debugMessageList, childError.UserMessage)
 		} else {
-			message = message + ": " + err.Error()
+			debugMessageList = append(debugMessageList, childError.DebugMessage)
 		}
+	} else {
+		debugMessageList = append(debugMessageList, err.Error())
 	}
 
-	resourceError.DebugMessage = message
+	e.DebugMessage = strings.Join(debugMessageList, ": ")
```

Comment:
We might need some tests for this thing.  I don't yet see how this list could ever have more than one entry in it.